File: module3/taskC.cpp
// Copyright Boris Ermolovich ermolovich.boris@gmail.com
#include <iostream>
#include <vector>
#include <queue>
#include <algorithm>
#include <cmath>
#include <limits>

using namespace std;

struct Item {
    int num;
    int weight;
    int cost;

    Item(int n, int w, int c) : num(n), weight(w), cost(c) {}
};

struct Solution {
    int total_weight;
    int total_cost;
    vector<int> indexes;

    Solution(int tw, int tc, const vector<int>& idx) : total_weight(tw), total_cost(tc), indexes(idx) {}

    void output() const {
        cout << total_weight << " " << total_cost << endl;
        for (int index : indexes) {
            cout << index << endl;
        }
    }
};

class KnapsackProblem {
    struct Node {
        int level;
        int value;
        int weight;
        double bound;
        vector<int> selection;

        Node(int l, int v, int w, double b, const vector<int>& sel)
            : level(l), value(v), weight(w), bound(b), selection(sel) {}

        bool operator<(const Node& other) const {
            return bound < other.bound;
        }
    };

    vector<Item> items;
    int capacity;
    double precision;

public:
    KnapsackProblem(const vector<Item>& items, int capacity, double precision)
        : items(items), capacity(capacity), precision(precision) {}

    double bound(const Node& node, const vector<Item>& sorted_items) {
        if (node.weight > capacity) {
            return 0;
        }

        double result = node.value;
        int total_weight = node.weight;

        for (size_t i = node.level + 1; i < sorted_items.size(); ++i) {
            const Item& item = sorted_items[i];
            if (total_weight + item.weight <= capacity) {
                total_weight += item.weight;
                result += item.cost;
            } else {
                result += (capacity - total_weight) * (double)item.cost / item.weight;
                break;
            }
        }

        return result;
    }

    Solution approximate_solution() {
        int max_cost = 0;
        for (const auto& item : items) {
            max_cost = max(max_cost, item.cost);
        }

        double scale = precision * max_cost / (items.size() * (1 + precision));

        vector<Item> scaled_items;
        for (const auto& item : items) {
            scaled_items.emplace_back(item.num, item.weight, (int)(item.cost / scale));
        }

        sort(scaled_items.begin(), scaled_items.end(), [](const Item& a, const Item& b) {
            return (double)a.cost / a.weight > (double)b.cost / b.weight;
        });

        priority_queue<Node> queue;
        Node root(-1, 0, 0, 0.0, {});
        root.bound = bound(root, scaled_items);
        queue.push(root);

        int max_value = 0;
        vector<int> best_selection;

        while (!queue.empty()) {
            Node node = queue.top();
            queue.pop();

            if (node.bound <= max_value) {
                continue;
            }

            if (node.level + 1 < (int)scaled_items.size()) {
                const Item& next_item = scaled_items[node.level + 1];

                Node node_with_item(
                    node.level + 1,
                    node.value + next_item.cost,
                    node.weight + next_item.weight,
                    0.0,
                    node.selection
                );
                node_with_item.selection.push_back(1);

                if (node_with_item.weight <= capacity) {
                    if (node_with_item.value > max_value) {
                        max_value = node_with_item.value;
                        best_selection = node_with_item.selection;
                    }
                    node_with_item.bound = bound(node_with_item, scaled_items);
                    if (node_with_item.bound > max_value) {
                        queue.push(node_with_item);
                    }
                }

                Node node_without_item(
                    node.level + 1,
                    node.value,
                    node.weight,
                    0.0,
                    node.selection
                );
                node_without_item.selection.push_back(0);

                node_without_item.bound = bound(node_without_item, scaled_items);
                if (node_without_item.bound > max_value) {
                    queue.push(node_without_item);
                }
            }
        }

        vector<int> selected_indices;
        for (size_t i = 0; i < best_selection.size(); ++i) {
            if (best_selection[i] == 1) {
                selected_indices.push_back(scaled_items[i].num);
            }
        }

        int total_weight = 0;
        int total_cost = 0;
        for (int index : selected_indices) {
            total_weight += items[index - 1].weight;
            total_cost += items[index - 1].cost;
        }

        return Solution(total_weight, total_cost, selected_indices);
    }
};

int main() {
    double approximation_coefficient;
    cin >> approximation_coefficient;

    int max_weight;
    cin >> max_weight;

    vector<Item> items;
    int i = 1;
    int weight, cost;

    while (cin >> weight >> cost) {
        items.emplace_back(i, weight, cost);
        ++i;
    }

    KnapsackProblem knapsack(items, max_weight, approximation_coefficient);
    Solution solution = knapsack.approximate_solution();
    solution.output();

    return 0;
}

File: module3/taskC.py
# Copyright Boris Ermolovich ermolovich.boris@gmail.com
# приближенный рюкзак
import sys
from queue import PriorityQueue

class Item:
    def __init__(self, num: int, weight: int, cost: int):
        self.num = num
        self.weight = weight
        self.cost = cost


class Solution:
    def __init__(self, total_weight: int, total_cost: int, indexes: [int]):
        self.total_weight = total_weight
        self.total_cost = total_cost
        self.indexes = indexes

    def output(self, output_handler=None):
        if not output_handler:
            solution_to_console(self)
        else:
            output_handler(self)


def solution_to_console(solution: Solution):
    print(solution.total_weight, solution.total_cost)
    for index in sorted(solution.indexes):
        print(index)


class KnapsackProblem:

    def __init__(self, items: [Item], capacity: int, precision: float):
        self.items = items
        self.capacity = capacity
        self.precision = precision

    # узел для метода ветвей и границ
    class Node:
        def __init__(self, level, value, weight, bound, selection):
            self.level = level
            self.value = value
            self.weight = weight
            self.bound = bound
            self.selection = selection

        def __lt__(self, other):
            return self.bound > other.bound

    # расчет предельного значения
    def bound(self, node, sorted_items):
        if node.weight > self.capacity:
            return 0
        result = node.value
        total_weight = node.weight
        for i in range(node.level + 1, len(sorted_items)):
            item = sorted_items[i]
            if total_weight + item.weight <= self.capacity:
                total_weight += item.weight
                result += item.cost_scaled
            else:
                result += (self.capacity - total_weight) * item.cost_scaled / item.weight
                break
        return result

    # масштабирование + метод ветвей и границ
    def approximate_solution(self) -> Solution:
        max_cost = max(item.cost for item in self.items)
        scale = self.precision * max_cost / (len(self.items) * (1 + self.precision))

        for item in self.items:
            item.cost_scaled = int(item.cost // scale)

        sorted_items = sorted(self.items, key=lambda x: x.cost_scaled / x.weight if x.weight > 0 else float('inf'),
                              reverse=True)

        queue = PriorityQueue()
        root = self.Node(-1, 0, 0, 0, [])
        root.bound = self.bound(root, sorted_items)
        queue.put(root)

        max_value = 0
        best_selection = []

        while not queue.empty():
            node = queue.get()

            if node.bound <= max_value:
                continue

            if node.level + 1 < len(sorted_items):
                next_item = sorted_items[node.level + 1]

                node_with_item = self.Node(
                    level=node.level + 1,
                    value=node.value + next_item.cost_scaled,
                    weight=node.weight + next_item.weight,
                    bound=0,
                    selection=node.selection + [1]
                )
                if node_with_item.weight <= self.capacity:
                    if node_with_item.value > max_value:
                        max_value = node_with_item.value
                        best_selection = node_with_item.selection
                    node_with_item.bound = self.bound(node_with_item, sorted_items)
                    if node_with_item.bound > max_value:
                        queue.put(node_with_item)

                node_without_item = self.Node(
                    level=node.level + 1,
                    value=node.value,
                    weight=node.weight,
                    bound=0,
                    selection=node.selection + [0]
                )
                node_without_item.bound = self.bound(node_without_item, sorted_items)
                if node_without_item.bound > max_value:
                    queue.put(node_without_item)

        selected_indices = [
            self.items.index(sorted_items[i]) + 1 for i in range(len(best_selection))
            if best_selection[i] == 1
        ]
        total_weight = sum(self.items[i - 1].weight for i in selected_indices)
        total_cost = sum(self.items[i - 1].cost for i in selected_indices)

        return Solution(total_weight=total_weight, total_cost=total_cost, indexes=selected_indices)



def main():
    approximation_coefficient = float(sys.stdin.readline())
    sys.stdin.flush()
    max_weight = int(sys.stdin.readline())
    sys.stdin.flush()
    items = []
    i = 1
    for line in sys.stdin.read().splitlines():
        items.append(Item(i, int(line.split()[0]), int(line.split()[1])))
        i += 1
    knapsack = KnapsackProblem(items, max_weight, approximation_coefficient)
    solution = knapsack.approximate_solution()
    solution.output()


if __name__ == '__main__':
    main()